File: 25-1.rb
#!/usr/bin/env ruby

require 'numo/narray'

class Map
  def initialize(input)
    @height = input.size
    @width = input[0].size
    @map = Numo::UInt8.zeros(@height, @width)

    input.each_index do |row|
      input[row].each_index do |col|
        @map[row, col] = decode(input[row][col])
      end
    end
  end

  def decode(symbol)
    case symbol
    when '.'
      0
    when '>'
      1
    when 'v'
      2
    end
  end

  def encode(number)
    case number
    when 0
      '.'
    when 1
      '>'
    when 2
      'v'
    end
  end

  def step!
    changed = false

    new_map = Numo::UInt8.zeros(@height, @width)

    @height.times do |y|
      @width.times do |x|
        new_map[y, x] = 2 if @map[y, x] == 2

        next if @map[y, x] != 1

        if @map[y, (x + 1) % @width] != 0
          new_map[y, x] = 1
        else
          new_map[y, x] = 0
          new_map[y, (x + 1) % @width] = 1
          changed = true
        end
      end
    end

    @map = new_map
    new_map = Numo::UInt8.zeros(@height, @width)

    @width.times do |x|
      @height.times do |y|
        new_map[y, x] = 1 if @map[y, x] == 1

        next if @map[y, x] != 2

        if @map[(y + 1) % @height, x] != 0
          new_map[y, x] = 2
        else
          new_map[y, x] = 0
          new_map[(y + 1) % @height, x] = 2
          changed = true
        end
      end
    end

    @map = new_map
    changed
  end

  def to_s
    s = "<#{self.class}: (height: #{@height}, width: #{@width})\n"
    @height.times do |y|
      @width.times do |x|
        s += encode(@map[y, x])
      end
      s += "\n"
    end
    s += '>'
    s
  end

  def inspect
    to_s
  end
end

input = File.read('25.input').lines.map(&:strip).map { |l| l.split '' }

map = Map.new input

steps = 0
loop do
  steps += 1
  break unless map.step!
end

print steps, "\n"

File: 25-1.py
#!/usr/bin/env python

import numpy

class Map:
    def __init__(self, lines):
        self.height = len(lines)
        self.width = len(lines[0])
        self.map = numpy.zeros((self.height, self.width), dtype=numpy.uint)

        for row in range(self.height):
            for col in range(self.width):
                self.map[row, col] = self.decode(lines[row][col])

    def decode(self, symbol):
        if symbol == ".":
            return 0
        elif symbol == '>':
            return 1
        elif symbol == 'v':
            return 2

    def encode(self, number):
        if number == 0:
            return "."
        elif number == 1:
            return ">"
        elif number == 2:
            return "v"

    def step(self):
        changed = False

        new_map = numpy.zeros((self.height, self.width), dtype=numpy.uint)

        for y in range(self.height):
            for x in range(self.width):
                if self.map[y, x] == 2:
                    new_map[y, x] = 2

                if self.map[y, x] != 1:
                    continue

                if self.map[y, (x + 1) % self.width] != 0:
                    new_map[y, x] = 1
                else:
                    new_map[y, x] = 0
                    new_map[y, (x + 1) % self.width] = 1
                    changed = True

        self.map = new_map

        new_map = numpy.zeros((self.height, self.width), dtype=numpy.uint)

        for x in range(self.width):
            for y in range(self.height):
                if self.map[y, x] == 1:
                    new_map[y, x] = 1

                if self.map[y, x] != 2:
                    continue

                if self.map[(y + 1) % self.height, x] != 0:
                    new_map[y, x] = 2
                else:
                    new_map[y, x] = 0
                    new_map[(y + 1) % self.height, x] = 2
                    changed = True

        self.map = new_map
        return changed


    def __str__(self):
        s = f"<{self.__class__.__name__}: (height: {self.height}, width: {self.width})\n"
        for y in range(self.height):
            for x in range(self.width):
                s += self.encode(self.map[y, x])
            s += "\n"
        s += '>'
        return s


lines = [list(line.strip()) for line in open('25.input').readlines()]

map = Map(lines)

steps = 0

while True:
    steps += 1
    if not map.step():
        break

print(steps)
